python/ray/data/datasource/parquet_datasource.py

```diff
@@ -56,6 +57,64 @@ def _deregister_parquet_file_fragment_serialization():
     ray.util.deregister_serializer(ParquetFileFragment)
 
 
+# easier to test when isolated from retry logic
+def _deserialize_pieces(
+    serialized_pieces: str,
+) -> List["pyarrow._dataset.ParquetFileFragment"]:
+    from ray import cloudpickle
+
+    pieces: List["pyarrow._dataset.ParquetFileFragment"] = cloudpickle.loads(
+        serialized_pieces
+    )
+    return pieces
+
+
+# this retry helps when the upstream datasource is not able to handle
+# overloaded read request or failed with some retriable failures.
+# For example when reading data from HA hdfs service, hdfs might
+# lose connection for some unknown reason expecially when
+# simutaneously running many hyper parameter tuning jobs
+# with ray.data parallelism setting at high value like the default 200
+# Such connection failure can be restored with some waiting and retry.
+def _deserialize_pieces_with_retry(
+    serialized_pieces: str,
+) -> List["pyarrow._dataset.ParquetFileFragment"]:
+    min_interval = 0
+    final_exception = None
+    for i in range(FILE_READING_RETRY):
+        try:
+            return _deserialize_pieces(serialized_pieces)
+        except Exception as e:
+            import time
+            import random
+
+            retry_timing = "" if i == FILE_READING_RETRY - 1 else (f"Retry after {min_interval} sec. ")
```

Comment:
There is some lint failure, please fix the format following the suggestion in error message: https://buildkite.com/ray-project/ray-builders-pr/builds/35756#01816ba1-d3be-494b-b637-a18ef120630b

Reply:
done.